store/tikv/2pc.go

```diff
@@ -226,6 +262,296 @@ func newTwoPhaseCommitter(txn *tikvTxn, connID uint64) (*twoPhaseCommitter, erro
+		ord := bytes.Compare(iter1.Key(), iter2.Key())
+		if ord < 0 {
+			iter = iter1
+		} else if ord > 0 {
+			iter = iter2
+		} else {
+			iter = iter1
+			equal = true
+		}
+		return
+	}
+
+	if iter1.Valid() {
+		iter = iter1
+	}
+	if iter2.Valid() {
+		iter = iter2
+	}
+	return
+}
+
+// jobControl wraps all the things for safe concurrency.
+// It contains the task channel and result channel,
+// It also provides run worker and send task to channel methods.
+// Something special about it is that the worker count is not fixed.
+type jobControl struct {
+	workerCnt int
+	taskCh    chan *batchMutations
+	// errCh is used by the worker to report something wrong, and exited is
+	// used to notify workers to exit when something is wrong.
+	errCh  chan error
+	exited chan struct{}
+}
+
+func (j *jobControl) init() {
+	j.errCh = make(chan error, 1) // nonblock for the first error
+	j.taskCh = make(chan *batchMutations)
+	j.exited = make(chan struct{})
+	j.workerCnt = 0
+}
+
+func (j *jobControl) sendBatchMutationToWorker(bo *Backoffer, b *batchMutations, c *twoPhaseCommitter, action twoPhaseCommitAction) error {
+	if b.mutations.len() == 0 {
+		return nil
+	}
+	// Try for the first time, nonblock sending the task to the channel.
+	select {
+	case j.taskCh <- b:
+		return nil
+	default:
+		// Create a new worker if necessary.
+		if j.workerCnt < 32 {
+			j.workerCnt++
+			go j.runWorker(bo, c, action)
+		}
+	}
+
+	// Block sending the task to the worker.
+	select {
+	case j.taskCh <- b:
+	case err := <-j.errCh:
+		return err
+	}
+	return nil
+}
+
+func (j *jobControl) runWorker(bo *Backoffer, c *twoPhaseCommitter, action twoPhaseCommitAction) {
+	var cancel context.CancelFunc
+	bo, cancel = bo.Fork()
+	defer cancel()
+	var err error
+	for batch := range j.taskCh {
+		err = action.handleSingleBatch(c, bo, batch)
+		if err != nil {
+			break
+		}
+	}
+	select {
+	case j.errCh <- err:
+	case <-j.exited:
+	}
+}
+
+func (j *jobControl) wait(alreadyFail error) error {
+	defer close(j.exited)
+
+	// Notify all workers task send done and wait for them to finish.
+	close(j.taskCh)
+
+	// On fail, exit early, close exited to avoid goroutine leak.
+	if alreadyFail != nil {
+		return alreadyFail
+	}
+
+	// On success, wait for all the worker to finish.
+	for i := 0; i < j.workerCnt; i++ {
+		err := <-j.errCh
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func do(bo *Backoffer, c *twoPhaseCommitter, action twoPhaseCommitAction) error {
+	txn := c.txn
+	iter, err := txn.us.GetMemBuffer().Iter(nil, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer iter.Close()
+	l := lockIter{lockKeys: txn.lockKeys}
+	bi := batchIter{iter: iter, lockIter: &l}
+
+	// Take the first batch.
+	batch, noMore, err := c.takeBatchMutationFromIterator(bo, action, &bi)
+	if err != nil || batch.mutations.len() == 0 {
+		return err
+	}
+	// Set the primary key.
+	batch.isPrimary = true
+	c.once.Do(func() {
+		c.primaryKey = batch.mutations.keys[0]
```

Comment:
The first key may not be the primary key.

Reply:
Yes, so this PR only handle the optimistic transaction case, the pessimistic is not handled yet.